File: app/src/main/java/com/example/pokedex/ui/detailpokemon/DetailPokemonFragment.java
package com.example.pokedex.ui.detailpokemon;

import static android.content.ContentValues.TAG;

import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.fragment.app.Fragment;
import androidx.lifecycle.ViewModelProvider;
import androidx.navigation.Navigation;

import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.bumptech.glide.Glide;
import com.example.pokedex.R;
import com.example.pokedex.databinding.FragmentDetailPokemonBinding;
import com.example.pokedex.utils.BaseDialog;
import com.example.pokedex.utils.Constant;
import com.example.pokedex.utils.Helper;
import com.google.android.material.snackbar.Snackbar;

import java.util.Random;

import dagger.hilt.android.AndroidEntryPoint;

@AndroidEntryPoint
public class DetailPokemonFragment extends Fragment {

  private FragmentDetailPokemonBinding binding;
  private DetailPokemonViewModel viewModel;
  private BaseDialog dialog;

  public DetailPokemonFragment() {
    // Required empty public constructor
  }

  @Override
  public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    // Inflate the layout for this fragment
    viewModel = new ViewModelProvider(this).get(DetailPokemonViewModel.class);
    binding = FragmentDetailPokemonBinding.inflate(inflater, container, false);
    return binding.getRoot();
  }

  @Override
  public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
    super.onViewCreated(view, savedInstanceState);
    int id = requireArguments().getInt(Constant.EXTRA_POKEMON_ID);
    Log.d(TAG, "onViewCreated: " + id);
    showDetailPokemon(id);
    binding.icBack.setOnClickListener(view1 -> {
      Navigation.findNavController(requireView()).navigateUp();
    });
  }

  private void caughtPokemon() {
    viewModel.getDetailPokemon().observe(getViewLifecycleOwner(), detailPokemonEntityResource -> {
      if (detailPokemonEntityResource != null) {
        switch (detailPokemonEntityResource.status) {
          case SUCCESS:
            if (detailPokemonEntityResource.data != null) {
              boolean state = detailPokemonEntityResource.data.isCaught();
              if (state) {
                binding.tvCatchPokemon.setText("Release!");
              } else {
                binding.tvCatchPokemon.setText("Catch!");
              }
              Log.d(TAG, "caughtPokemon: " + detailPokemonEntityResource.data.isCaught());
              binding.ivCatch.setOnClickListener(view -> {
                Random random = new Random();
                int randomNumber = random.nextInt(3 - 1) + 1;
                if (!state) {
                  if (randomNumber == 1) {
                    dialog = new BaseDialog(requireContext(), R.layout.fragment_bottom_dialog);
                    dialog.show();
                    dialog.updateText(new BaseDialog.DialogCallback() {
                      @Override public void onDialogShow(@NonNull String text) {
                        viewModel.setCaughtPokemon(text);
                        dialog.dismiss();
                        Snackbar.make(view, text + " has been added to pokemon list",
                            Snackbar.LENGTH_SHORT).show();
                      }
                    });
                  } else {
                    Snackbar.make(view, "You Failed to Catch a Pokemon", Snackbar.LENGTH_SHORT)
                        .show();
                  }
                } else {
                  Snackbar.make(view, "You Released a Pokemon", Snackbar.LENGTH_SHORT).show();
                  viewModel.setCaughtPokemon(" ");
                }
              });
            }
        }
      }
    });
  }

  private void showDetailPokemon(int id) {
    viewModel.setPokemonId(id);
    viewModel.getDetailPokemon().observe(getViewLifecycleOwner(), detailPokemonResponse -> {
      if (detailPokemonResponse != null) {
        switch (detailPokemonResponse.status) {
          case LOADING:
            binding.progressBar.setVisibility(View.VISIBLE);
            break;
          case SUCCESS:
            binding.progressBar.setVisibility(View.GONE);
            Glide.with(requireContext())
                .load("https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/"
                    + id
                    + ".png")
                .into(binding.ivPokemon);
            binding.tvPokemonName.setText(detailPokemonResponse.data.getName());
            binding.tvPokemonHeight.setText(
                Helper.getHeight(detailPokemonResponse.data.getHeight()));
            binding.tvPokemonWeight.setText(
                Helper.getWeight(detailPokemonResponse.data.getWeight()));
            binding.tvPokemonId.setText(Helper.idConverter(id));
            break;
          case ERROR:
            binding.progressBar.setVisibility(View.GONE);
            Toast.makeText(requireContext(), "Error occurred", Toast.LENGTH_SHORT).show();
        }
      }
    });
    caughtPokemon();
  }
}

File: app/src/main/java/com/example/pokedex/ui/detailpokemon/DetailPokemonFragment.kt
package com.example.pokedex.ui.detailpokemon

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.fragment.app.Fragment
import androidx.fragment.app.viewModels
import androidx.navigation.Navigation
import com.example.domain.model.DetailPokemonModel
import com.example.domain.utils.Resource
import com.example.pokedex.R.layout
import com.example.pokedex.databinding.FragmentDetailPokemonBinding
import com.example.pokedex.utils.BaseDialog
import com.example.pokedex.utils.BaseDialog.DialogCallback
import com.example.pokedex.utils.Constant
import com.example.pokedex.utils.Helper.getHeight
import com.example.pokedex.utils.Helper.getWeight
import com.example.pokedex.utils.Helper.idConverter
import com.example.pokedex.utils.ImageUtil
import com.google.android.material.snackbar.Snackbar
import dagger.hilt.android.AndroidEntryPoint

@AndroidEntryPoint
class DetailPokemonFragment : Fragment() {
  private var _binding: FragmentDetailPokemonBinding? = null
  private val binding get() = _binding!!
  val viewModel: DetailPokemonViewModel by viewModels()
  private var dialog: BaseDialog? = null
  override fun onCreateView(
    inflater: LayoutInflater, container: ViewGroup?,
    savedInstanceState: Bundle?
  ): View? {
    _binding = FragmentDetailPokemonBinding.inflate(inflater, container, false)
    return binding.root
  }

  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
    super.onViewCreated(view, savedInstanceState)
    val id = requireArguments().getInt(Constant.EXTRA_POKEMON_ID)
    showDetailPokemon(id)
    binding.icBack.setOnClickListener {
      Navigation.findNavController(requireView()).navigateUp()
    }
  }

  private fun caughtPokemon() {
    viewModel.detailPokemon.observe(viewLifecycleOwner) { detailPokemonEntityResource: Resource<DetailPokemonModel> ->
      when (detailPokemonEntityResource) {
        is Resource.Success ->
          if (detailPokemonEntityResource.data != null) {
            val state = detailPokemonEntityResource.data!!.isCaught
            if (state) {
              binding.tvCatchPokemon.text = "Release!"
            } else {
              binding.tvCatchPokemon.text = "Catch!"
            }
            var name = if (detailPokemonEntityResource.data!!.nickname == null
              || detailPokemonEntityResource.data!!.nickname == ""
            ) detailPokemonEntityResource.data!!.name else detailPokemonEntityResource.data!!.nickname!!
            binding.ivCatch.setOnClickListener {

              val randomCatch = viewModel.getRandomLogic
              if (!state) {
                if (randomCatch == 1) {
                  dialog = BaseDialog(requireContext(), layout.fragment_bottom_dialog)
                  dialog?.show()
                  dialog?.setNickname(object : DialogCallback {
                    override fun onDialogShow(text: String) {
                      dialog?.dismiss()
                      viewModel.setCaughtPokemon(text)
                      Snackbar.make(
                        requireView(), "$name has been added to pokemon list",
                        Snackbar.LENGTH_SHORT
                      ).show()
                      return
                    }
                  })

                } else {
                  Snackbar.make(
                    requireView(),
                    "You Failed to Catch a Pokemon",
                    Snackbar.LENGTH_SHORT
                  )
                    .show()
                }
              } else {
                viewModel.setCaughtPokemon("")
                Snackbar.make(requireView(), "$name has been released", Snackbar.LENGTH_SHORT)
                  .show()

              }
            }
          }
      }
    }
  }

  private fun showDetailPokemon(id: Int) {
    viewModel.setPokemonId(id)
    viewModel.detailPokemon.observe(viewLifecycleOwner) { detailPokemonResponse: Resource<DetailPokemonModel>? ->
      if (detailPokemonResponse != null) {
        when (detailPokemonResponse) {
          is Resource.Loading -> binding.progressBar.visibility = View.VISIBLE
          is Resource.Success -> {
            binding.progressBar.visibility = View.GONE
            ImageUtil.generateBackgroundPalette(
              requireContext(),
              "https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/$id.png",
              binding.ivPokemon,
              binding.actDetail
            )
            val name = if (detailPokemonResponse.data!!.nickname == null
              || detailPokemonResponse.data!!.nickname == ""
            ) detailPokemonResponse.data!!.name else detailPokemonResponse.data!!.nickname!!
            binding.tvPokemonName.text = name
            binding.tvPokemonHeight.text = getHeight(
              detailPokemonResponse.data!!.height.toDouble()
            )
            binding.tvPokemonWeight.text = getWeight(
              detailPokemonResponse.data!!.weight.toDouble()
            )
            binding.tvPokemonId.text = idConverter(id)
          }
          is Resource.Error -> {
            binding.progressBar.visibility = View.GONE
            Toast.makeText(requireContext(), "Error occurred", Toast.LENGTH_SHORT).show()
          }
        }
      }
    }
    caughtPokemon()
  }
}